File: core/types/automaton.ts
/**
 * Core domain types shared across services. These are wire-agnostic
 * representations; gRPC and REST layers translate to their respective
 * formats.
 */

import { Money } from "./money.js";
import { AutomatonId, PlanId, TaskId, ActionId, MemoryId } from "./ids.js";

export type ISO8601 = string;

export type LifecycleState =
  | "created"
  | "running"
  | "paused"
  | "suspended"
  | "replicating"
  | "terminated"
  | "archived";

export type RiskLevel = "low" | "medium" | "high" | "critical";

export interface Automaton {
  id: AutomatonId;
  name: string;
  genesisPrompt: string;
  parentId: AutomatonId | null;
  publicKey: string; // base64 Ed25519
  walletAddress: string;
  state: LifecycleState;
  createdAt: ISO8601;
  updatedAt: ISO8601;
  version: string;
  reputation: number; // 0..1
  baseCurrency: string;
  balance: Money;
  budget: Money;
  metadata: Record<string, string>;
}

export interface Goal {
  id: string;
  description: string;
  priority: number; // 0..100
  expectedRevenue: Money;
  estimatedCost: Money;
  probability: number; // 0..1
  status: "pending" | "active" | "completed" | "failed" | "cancelled";
  createdAt: ISO8601;
  completedAt?: ISO8601;
}

export interface Plan {
  id: PlanId;
  automatonId: AutomatonId;
  goalId: string;
  steps: PlanStep[];
  estimatedCost: Money;
  expectedRevenue: Money;
  probability: number;
  createdAt: ISO8601;
  status: "draft" | "approved" | "executing" | "succeeded" | "failed" | "cancelled";
}

export interface PlanStep {
  index: number;
  kind: string; // tool | llm | external
  description: string;
  estimatedCost: Money;
  dependsOn: number[];
}

export interface Task {
  id: TaskId;
  automatonId: AutomatonId;
  kind: string;
  payload: Record<string, unknown>;
  budget: Money;
  deadline?: ISO8601;
  status:
    | "queued"
    | "in_progress"
    | "awaiting_payment"
    | "succeeded"
    | "failed"
    | "expired"
    | "cancelled";
  result?: Record<string, unknown>;
  createdAt: ISO8601;
  updatedAt: ISO8601;
}

export interface Action {
  id: ActionId;
  taskId: TaskId;
  planId: PlanId;
  toolName: string;
  arguments: Record<string, unknown>;
  risk: RiskLevel;
  costEstimate: Money;
  policyDecision: PolicyDecision;
  startedAt: ISO8601;
  completedAt?: ISO8601;
  result?: unknown;
  error?: string;
}

export type PolicyVerdict = "allow" | "deny" | "require_approval";

export interface PolicyDecision {
  verdict: PolicyVerdict;
  reason: string;
  evaluatedAt: ISO8601;
  evaluator: string; // "constitution@v1" | "rbac" | "abac" | ...
  citations: string[]; // e.g. ["constitution:law:1", "policy:budget"]
  expiresAt?: ISO8601;
}

export interface MemoryEntry {
  id: MemoryId;
  automatonId: AutomatonId;
  layer: MemoryLayer;
  content: string;
  embedding?: number[]; // base64-encoded float32 vector
  importance: number; // 0..1
  ttl?: number; // seconds
  createdAt: ISO8601;
  updatedAt: ISO8601;
  tags: string[];
}

export type MemoryLayer =
  | "working"
  | "short_term"
  | "long_term"
  | "semantic"
  | "procedural"
  | "financial"
  | "operational"
  | "code_history"
  | "decision_history"
  | "relationship";

export interface TreasuryEntry {
  id: string;
  automatonId: AutomatonId;
  kind: "credit" | "debit";
  amount: Money;
  category: string; // "llm_inference" | "compute" | "revenue:api" | ...
  refType?: string;
  refId?: string;
  occurredAt: ISO8601;
  memo?: string;
}

export interface Tool {
  name: string;
  version: string;
  description: string;
  capabilities: string[];
  risk: RiskLevel;
  cost: Money;
  rateLimit?: { perMinute: number; perHour: number; perDay: number };
  sandbox: "none" | "process" | "container" | "microvm";
  schema: unknown; // JSON schema for arguments
}

export interface HealthReport {
  status: "healthy" | "degraded" | "unhealthy";
  components: Record<string, ComponentHealth>;
  checkedAt: ISO8601;
}

export interface ComponentHealth {
  status: "up" | "down" | "degraded";
  latencyMs?: number;
  message?: string;
}

File: core/types/automaton.py
"""
Pydantic models for the core domain. These are used at service boundaries
(API, gRPC, DB serialization). Internally, the runtime uses lighter dataclasses
for hot paths.
"""
from __future__ import annotations

import re
from datetime import datetime
from enum import Enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from .identifiers import (
    ActionId,
    AutomatonId,
    EventId,
    MemoryId,
    PlanId,
    TaskId,
    new_automaton_id,
)
from .money import Money


class LifecycleState(str, Enum):
    CREATED = "created"
    RUNNING = "running"
    PAUSED = "paused"
    SUSPENDED = "suspended"
    REPLICATING = "replicating"
    TERMINATED = "terminated"
    ARCHIVED = "archived"


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class PolicyVerdict(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    REQUIRE_APPROVAL = "require_approval"


class MemoryLayer(str, Enum):
    WORKING = "working"
    SHORT_TERM = "short_term"
    LONG_TERM = "long_term"
    SEMANTIC = "semantic"
    PROCEDURAL = "procedural"
    FINANCIAL = "financial"
    OPERATIONAL = "operational"
    CODE_HISTORY = "code_history"
    DECISION_HISTORY = "decision_history"
    RELATIONSHIP = "relationship"


class _Base(BaseModel):
    model_config = ConfigDict(
        extra="forbid", frozen=False, populate_by_name=True, arbitrary_types_allowed=True
    )


class Automaton(_Base):
    id: AutomatonId = Field(default_factory=new_automaton_id)
    name: str
    genesis_prompt: str
    parent_id: Optional[AutomatonId] = None
    public_key: str  # base64 Ed25519
    wallet_address: str
    state: LifecycleState = LifecycleState.CREATED
    created_at: datetime
    updated_at: datetime
    version: str = "0.1.0"
    reputation: float = Field(default=0.5, ge=0.0, le=1.0)
    base_currency: str = "USDC"
    balance: Money = Field(default_factory=lambda: Money.zero())
    budget: Money = Field(default_factory=lambda: Money.zero())
    metadata: dict[str, str] = Field(default_factory=dict)


class Goal(_Base):
    id: str
    description: str
    priority: int = Field(ge=0, le=100, default=50)
    expected_revenue: Money
    estimated_cost: Money
    probability: float = Field(ge=0.0, le=1.0, default=0.5)
    status: Literal["pending", "active", "completed", "failed", "cancelled"] = "pending"
    created_at: datetime
    completed_at: Optional[datetime] = None


class PlanStep(_Base):
    index: int
    kind: str  # tool | llm | external
    # description is intentionally permissive: a string for free-form steps
    # (e.g. an internal reflection note) or a dict of structured arguments
    # for a tool invocation. The runtime executor branches on the type.
    description: str | dict[str, Any]
    estimated_cost: Money
    risk: RiskLevel = RiskLevel.LOW
    depends_on: list[int] = Field(default_factory=list)


class Plan(_Base):
    id: PlanId
    automaton_id: AutomatonId
    goal_id: str
    steps: list[PlanStep]
    estimated_cost: Money
    expected_revenue: Money
    probability: float
    created_at: datetime
    status: Literal[
        "draft", "approved", "executing", "succeeded", "failed", "cancelled"
    ] = "draft"


class Task(_Base):
    id: TaskId
    automaton_id: AutomatonId
    kind: str
    payload: dict[str, Any]
    budget: Money
    deadline: Optional[datetime] = None
    status: Literal[
        "queued",
        "in_progress",
        "awaiting_payment",
        "succeeded",
        "failed",
        "expired",
        "cancelled",
    ] = "queued"
    result: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class PolicyDecision(_Base):
    verdict: PolicyVerdict
    reason: str
    evaluated_at: datetime
    evaluator: str
    citations: list[str] = Field(default_factory=list)
    expires_at: Optional[datetime] = None


class Action(_Base):
    id: ActionId
    task_id: TaskId
    plan_id: PlanId
    tool_name: str
    arguments: dict[str, Any]
    risk: RiskLevel
    cost_estimate: Money
    policy_decision: PolicyDecision
    started_at: datetime
    completed_at: Optional[datetime] = None
    result: Optional[Any] = None
    error: Optional[str] = None


class MemoryEntry(_Base):
    id: MemoryId
    automaton_id: AutomatonId
    layer: MemoryLayer
    content: str
    embedding: Optional[list[float]] = None
    importance: float = Field(ge=0.0, le=1.0, default=0.5)
    ttl: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    tags: list[str] = Field(default_factory=list)


class TreasuryEntry(_Base):
    id: str
    automaton_id: AutomatonId
    kind: Literal["credit", "debit"]
    amount: Money
    category: str
    ref_type: Optional[str] = None
    ref_id: Optional[str] = None
    occurred_at: datetime
    memo: Optional[str] = None


class ToolSpec(_Base):
    name: str
    version: str
    description: str
    capabilities: list[str]
    risk: RiskLevel
    cost: Money
    rate_limit: Optional[dict[str, int]] = None
    sandbox: Literal["none", "process", "container", "microvm"] = "process"
    schema_: dict[str, Any] = Field(default_factory=dict, alias="schema")

    @field_validator("name")
    @classmethod
    def _name(cls, v: str) -> str:
        # Allow dotted tool names like "browser.click" — alphanumeric, underscores, dots.
        if not re.match(r"^[A-Za-z0-9_]+(?:\.[A-Za-z0-9_]+)*$", v):
            raise ValueError("tool name must be alphanumeric/underscore/dot")
        return v


class ComponentHealth(_Base):
    status: Literal["up", "down", "degraded"]
    latency_ms: Optional[float] = None
    message: Optional[str] = None


class HealthReport(_Base):
    status: Literal["healthy", "degraded", "unhealthy"]
    components: dict[str, ComponentHealth]
    checked_at: datetime
